### PR TITLE
Add support for SHA256/192

### PR DESCRIPTION
**Pull request by fergus-dall**
This is defined in NIST SP 800-208 as the truncation to 192 bits of SHA256. Unlike other truncated hashes in the SHA2 suite, this variant doesn't have a different initial state, it is just a pure truncation of the output.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/recipes/30-test_evp_data/evpmd_sha.txt [hunk 1]
```diff
@@ -58,6 +58,22 @@ Input = "a"
...
+Digest = SHA256-192
+Input = "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
+Output = 248d6a61d20638b8e5c026930c3e6039a33ce45964ff2167
+
+Digest = SHA256-192
+Input = "a"
+Ncopy = 288
+Count = 3472
+Input = "a"
+Ncopy = 64
```
Comment (paulidale) [inline test/recipes/30-test_evp_data/evpmd_sha.txt:74]: Duplication of Ncopy and Input??

Comment (fergus-dall) [reply test/recipes/30-test_evp_data/evpmd_sha.txt:74]: Test input is built up cumulatively, so the extra line add another 64 `a`s to the end of the test input (for a total of exactly 1,000,000)

The same input is used for SHA256 above, I just copied those tests

**Review by paulidale — APPROVED** [review]

File: providers/implementations/digests/sha2_prov.c [hunk 2]
```diff
@@ -72,6 +72,11 @@ IMPLEMENT_digest_functions(sha256, SHA256_CTX,
                            SHA256_CBLOCK, SHA256_DIGEST_LENGTH, SHA2_FLAGS,
                            SHA256_Init, SHA256_Update, SHA256_Final)
 
+/* ossl_sha256_192_functions */
+IMPLEMENT_digest_functions(sha256_192, SHA256_CTX,
+                           SHA256_CBLOCK, SHA256_192_DIGEST_LENGTH, SHA2_FLAGS,
+                           SHA256_Init, SHA256_Update, SHA256_Final)
+
```
Comment (t8m) [inline providers/implementations/digests/sha2_prov.c:79]: Should it be inside `# ifndef FIPS_MODULE` ?

We do not make the implementation available from the fips provider so we should not build it there.

Comment (fergus-dall) [reply providers/implementations/digests/sha2_prov.c]: Done

File: providers/implementations/digests/sha2_prov.c [hunk 3]
```diff
@@ -71,7 +71,12 @@ IMPLEMENT_digest_functions(sha224, SHA256_CTX,
 IMPLEMENT_digest_functions(sha256, SHA256_CTX,
                            SHA256_CBLOCK, SHA256_DIGEST_LENGTH, SHA2_FLAGS,
                            SHA256_Init, SHA256_Update, SHA256_Final)
-
+# ifndef FIPS_MODULE
+/* ossl_sha256_192_functions */
+IMPLEMENT_digest_functions(sha256_192, SHA256_CTX,
+                           SHA256_CBLOCK, SHA256_192_DIGEST_LENGTH, SHA2_FLAGS,
+                           sha256_192_init, SHA256_Update, SHA256_Final)
+#endif
```
Comment (t8m) [inline providers/implementations/digests/sha2_prov.c:79]: Missing space after `#`

File: providers/implementations/digests/sha2_prov.c [hunk 4]
```diff
@@ -71,7 +71,12 @@ IMPLEMENT_digest_functions(sha224, SHA256_CTX,
 IMPLEMENT_digest_functions(sha256, SHA256_CTX,
                            SHA256_CBLOCK, SHA256_DIGEST_LENGTH, SHA2_FLAGS,
                            SHA256_Init, SHA256_Update, SHA256_Final)
-
+# ifndef FIPS_MODULE
```
Comment (t8m) [inline providers/implementations/digests/sha2_prov.c:74]: The space between `#` and `ifndef` should be removed. Sorry, for misleading you.

File: crypto/sha/sha256.c [hunk 5]
```diff
@@ -53,6 +54,13 @@ int SHA256_Init(SHA256_CTX *c)
     return 1;
 }
 
+int sha256_192_init(SHA256_CTX *c)
```
Comment (paulidale) [inline crypto/sha/sha256.c:57]: This should be named `ossl_sha256_192_init`
We've adopted a naming convention whereby all exported but not public symbols start `ossl_`

File: test/recipes/30-test_evp_data/evpmac_common.txt [hunk 6]
```diff
@@ -67,6 +67,29 @@ BlockSize = 64
 
 Title = SHA2
 
+Availablein = default
+MAC = HMAC
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evpmac_common.txt:71]: Please include a comment about where the tests have come from
Do the same with test file below.

File: CHANGES.md [hunk 7]
```diff
@@ -279,6 +279,10 @@ OpenSSL 3.2
 
    *Čestmír Kalina*
 
+ * Added SHA256/192 algorithm support.
+
+   *Fergus Dall*
+
```
Comment (levitte) [inline CHANGES.md:285]: New entries are added at the top of the section, not the bottom

**Review by levitte — CHANGES_REQUESTED** [review]

File: crypto/sha/sha256.c (hunk 5 above)
Comment (fergus-dall) [reply crypto/sha/sha256.c:57]: Just to be clear, the similar functions in `include/crypto/sha.h` predate this convention?

File: test/recipes/30-test_evp_data/evpmac_common.txt (hunk 6 above)
Comment (fergus-dall) [reply test/recipes/30-test_evp_data/evpmac_common.txt:71]: I computed these with a python script, if you're hoping for a more official source AFAIK there isn't one

Likewise the tests in `evpmd_sha.txt` are just the SHA256 tests with the hash truncated

File: crypto/sha/sha256.c (hunk 5 above)
Comment (paulidale) [reply crypto/sha/sha256.c]: Yes.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the contribution.

**Pull request closed without merging**